kv/kvclient/kvcoord: TestPriorityRatchetOnAbortOrPush failed

kv/kvclient/kvcoord.TestPriorityRatchetOnAbortOrPush [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3276090&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3276090&tab=artifacts#/) on master @ [62ec88c61edcaa023a579199cc5b43d3ee951cef](https://github.com/cockroachdb/cockroach/commits/62ec88c61edcaa023a579199cc5b43d3ee951cef):


```
=== RUN   TestPriorityRatchetOnAbortOrPush
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestPriorityRatchetOnAbortOrPush351144054
    test_log_scope.go:74: use -show-logs to present logs inline
    txn_test.go:247: "read=false": expected priority on retry to ratchet to 2147483646; got 713736
    panic.go:613: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestPriorityRatchetOnAbortOrPush351144054
--- FAIL: TestPriorityRatchetOnAbortOrPush (1.08s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestPriorityRatchetOnAbortOrPush PKG=./pkg/kv/kvclient/kvcoord TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
/cc @cockroachdb/kv spencerkimball
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPriorityRatchetOnAbortOrPush.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvclient/kvcoord.TestPriorityRatchetOnAbortOrPush [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3345356&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3345356&tab=artifacts#/) on master @ [46cef2c6f0b36ba2f7d551c8ab017832c1b9d592](https://github.com/cockroachdb/cockroach/commits/46cef2c6f0b36ba2f7d551c8ab017832c1b9d592):


```
=== RUN   TestPriorityRatchetOnAbortOrPush
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestPriorityRatchetOnAbortOrPush179083822
    test_log_scope.go:74: use -show-logs to present logs inline
    txn_test.go:247: "read=false": expected priority on retry to ratchet to 2147483646; got 234506
    panic.go:613: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestPriorityRatchetOnAbortOrPush179083822
--- FAIL: TestPriorityRatchetOnAbortOrPush (1.09s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestPriorityRatchetOnAbortOrPush PKG=./pkg/kv/kvclient/kvcoord TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
/cc @cockroachdb/kv spencerkimball
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPriorityRatchetOnAbortOrPush.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvclient/kvcoord/txn_test.go
+++ b/pkg/kv/kvclient/kvcoord/txn_test.go
@@ -202,7 +202,17 @@ func TestLostUpdate(t *testing.T) {
 func TestPriorityRatchetOnAbortOrPush(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
-	s := createTestDB(t)
+	s := createTestDBWithKnobs(t, &kvserver.StoreTestingKnobs{
+		TestingRequestFilter: func(_ context.Context, ba roachpb.BatchRequest) *roachpb.Error {
+			// Reject transaction heartbeats, which can make the test flaky when they
+			// detect an aborted transaction before the Get operation does. See #68584
+			// for an explanation.
+			if ba.IsSingleHeartbeatTxnRequest() {
+				return roachpb.NewErrorf("rejected")
+			}
+			return nil
+		},
+	})
 	defer s.Stop()
 
 	pushByReading := func(key roachpb.Key) {

